src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -65178,5 +65178,35 @@ static void F3(int? z, int? w)
             var comp = CreateCompilationWithIndexAndRange(source, options: WithNonNullTypesTrue());
             comp.VerifyDiagnostics();
         }
+
+        [WorkItem(31770, "https://github.com/dotnet/roslyn/issues/31770")]
+        [Fact]
+        public void UserDefinedConversion_NestedNullability()
+        {
+            var source =
+@"class A<T> { }
+class B
+{
+    public static implicit operator B(A<object> a) => throw null;
+}
+class Program
+{
+    static void F(B b) { }
+    static void Main()
+    {
+        A<object?> a = new A<object?>();
+        B b = a; // 1
+        F(a); // 2
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (12,15): warning CS8619: Nullability of reference types in value of type 'A<object?>' doesn't match target type 'A<object>'.
```

Comment:
Is there any way we can make these errors more informative? It's relatively obvious in this simple example, but there's no mention in this error about the difference being part of the conversion, which in a more complicated scenario would be in another section of the code written by a different person.

Reply:
Logged https://github.com/dotnet/roslyn/issues/31798.

Reply:
Logged [https://github.com/dotnet/roslyn/issues/31798](https://github.com/dotnet/roslyn/issues/31798).

---
In reply to: [241617155](https://github.com/dotnet/roslyn/pull/31771#discussion_r241617155) [](ancestors = 241617155)